files/mkafka.c

```diff
@@ -0,0 +1,293 @@
+/* File mailbox.c */
+/* Define system calls to the mailbox */
+
+#include "mkafkalib.h"
+#include <minix/syslib.h>
+#include "glo.h"
+
+
+static mkafka_t *mkafka;
+static int mutex;
+
+
+/*
+ * Used for debugging purposes
+ * Print all messages which are currently in the mailbox
+
+int print_all_messages()
+{
+
+    int i;
+    message_t *message_ptr = mailbox->head;
+
+    for (i = 1; i <= mailbox->number_of_messages; i++)
+    {
+         message_ptr = message_ptr -> next;
+
+         pid_node_t *pids = message_ptr-> recipients;
+         char *message = message_ptr -> message;
+
+         //printf("**Message number %d\n", i);
+         //printf("**Message content %s\n", message);
+         //printf("**Recipients: ");
+
+         pids = pids -> next;
+
+         while(pids->pid != -1)
+         {
+             printf(" %d, ", pids->pid);
+             pids = pids -> next;
+         }
+         printf("\n");
+    }
+
+    return 0;
+ }
+*/
+
+/* initialized mkafka  */
+int init_mkafka (){
+    mkafka = malloc (sizeof(mkafka_t));
+    mkafka->head = NULL;
+    mkafka->num_topics = 0; 
+
+    mutex = 0;
+    return OK;
+}
+
+
+/* tbd */
+int remove_topic(char *id){
+    return 1;
+}
+
+/* Used to create a new topic
+ * Must be space
+ * id must be unique
+ */
+topic_t* add_topic(int id){
+    if (mkafka->num_topics < MAX_TOPICS){
+        topic_t *tp = mkafka->head;
+        if ( tp == NULL ){
+            //no topics yet
+            tp = malloc(sizeof(topic_t));
+            tp->id = id; 
+            tp->num_messages = 0; 
+            tp->next = NULL;
+            tp->head = NULL;
+            tp->tail = NULL;
+            mkafka->head = tp; 
+            mkafka->num_topics = 1;
+        }
+        else {
+            //find the next avalible slot,
+            while(tp->next != NULL) {
+               tp = tp->next;
+            }
+            topic_t *new_topic = malloc(sizeof(topic_t));
+            tp->next = new_topic;
+            new_topic->next = NULL;
+            new_topic->id = id;
+            new_topic->head = NULL; 
+            new_topic->tail = NULL; 
+            new_topic->num_messages = 0;
+            tp = new_topic;
+            mkafka->num_topics += 1;
+        }
+        //set up the listeners
+        for (int i = 0; i < MAX_GROUPS; i++){
+            tp->groups[i] = tp->head;
+        }
+        return tp;
+    }
+    else {
+        printf ("Too Many topics\n");
+        return NULL;
+    }
+}
+
+
+/* Creates mailbox if it is missing
+ * Add message to mailbox (if mailbox is not full)
+ * Returns OK if message was successfully added
+ * Returns ERROR if mailbox is full
+ */
+int push_to_topic()
+{
+    while (mutex==1);
```

Comment:
bless the cpu

Reply:
I don't actually know how to make this not shit, i'm putting you in charge :).